policy/filter.go

```diff
@@ -0,0 +1,155 @@
+		return
+	}
+
+	// buffer both of these channels, to prevent the goroutines below from
+	// blocking on send between checks of ctx.Done()
+	upstreamPolicyCh := make(chan IDMessage, 1)
+	filterUpdateCh := make(chan struct{}, 1)
+
+	// create separate error channels just for augmenting the log messages
+	policyErrCh := make(chan error, 1)
+	filterErrCh := make(chan error, 1)
+
+	go fs.upstreamSource.MonitorIDs(ctx, MonitorIDsReq{
+		ErrCh:    policyErrCh,
+		ResultCh: upstreamPolicyCh,
+	})
+	go fs.policyFilter.MonitorFilterUpdates(ctx, MonitorFilterRequest{
+		ErrCh:    filterErrCh,
+		UpdateCh: filterUpdateCh,
+	})
+
+	// keep track of the previous policyIDs, in case the filter updates
+	var policyIDs []PolicyID
+
+	// don't emit policy IDs until both the filter and the upstream policy
+	// source have sent their first update
+	haveFirstPolicies, haveFirstFilter := false, false
+
+	for {
+		select {
+		case <-ctx.Done():
+			fs.log.Trace("stopping file policy source ID monitor")
+			return
+
+		case err := <-policyErrCh:
+			if err == nil {
+				continue
+			}
+			req.ErrCh <- fmt.Errorf("error from upstream policy source monitor: %v", err)
+			continue
+
+		case err := <-filterErrCh:
+			if err == nil {
+				continue
+			}
+			req.ErrCh <- fmt.Errorf("error from policy filter monitor: %v", err)
+			continue
+
+		case newUpstreamIDs := <-upstreamPolicyCh:
+			policyIDs = newUpstreamIDs.IDs
+			haveFirstPolicies = true
+
+		case <-filterUpdateCh:
+			haveFirstFilter = true
+		}
+
+		if !haveFirstPolicies || !haveFirstFilter {
+			continue
+		}
+
+		newPolicyIDs := fs.policyFilter.FilterPolicies(policyIDs)
+		fs.log.Debug("filtered policies", "original_len", len(policyIDs), "filtered_len", len(newPolicyIDs))
+		req.ResultCh <- IDMessage{
+			IDs:    newPolicyIDs,
+			Source: fs.Name(),
+		}
+	}
+}
+
+// MonitorPolicy calls the same method on the configured policy.Source.
+// This method doesn't need to worry about the policy filter, because the policy.Manager
+// will close the context if the corresponding policy is removed.
+func (fs *FilteredSource) MonitorPolicy(ctx context.Context, req MonitorPolicyReq) {
+	fs.log.Trace("delegating MonitorPolicy", "policy_id", req.ID)
+	fs.upstreamSource.MonitorPolicy(ctx, req)
+}
+
+// Name satisfies the Name function of the policy.Source interface.
+// There needs to be a match between:
+// * the name reported by the policy source during MonitorIDs
+// * this method
+// * the key for the map of sources in the policy Manager
+// * and maybe other stuff as well.
+// Therefore, we'll continue to use the existing name.
+func (fs *FilteredSource) Name() SourceName {
+	return fs.upstreamSource.Name()
+}
+
+// ReloadIDsMonitor implements policy.Source by calling the appropriate
+// reload method on the underlying policy source and filter.
+func (fs *FilteredSource) ReloadIDsMonitor() {
+	fs.upstreamSource.ReloadIDsMonitor()
+	fs.policyFilter.ReloadFilterMonitor()
```

Comment:
Do we need to reset `haveFirstPolicies` and `haveFirstFilter` here? I'm not sure how they relate, but I'm wondering if it could cause the unfiltered policies to be sent on reload.

Reply:
We don't. Those two variables ensure that we have both members and policy responses available before performing the first set of assignments. The current reloading which does need broader work to improve, is a different starting point and has potential to change something like a TLS certificate only. If we were to reset these values, it would be possible policy changes are not processed because the pool membership has not changed, and therefore the blocking query not returned.

This potentially changes in the future when we enhance the reload capabilities; but that feels like a large piece of work outside this particular feature.